simde/x86/gfni.h

```diff
@@ -0,0 +1,765 @@
+   0x0b, 0x28, 0x2f, 0xa3, 0xda, 0xd4, 0xe4, 0x0f, 0xa9, 0x27, 0x53, 0x04, 0x1b, 0xfc, 0xac, 0xe6,
+   0x7a, 0x07, 0xae, 0x63, 0xc5, 0xdb, 0xe2, 0xea, 0x94, 0x8b, 0xc4, 0xd5, 0x9d, 0xf8, 0x90, 0x6b,
+   0xb1, 0x0d, 0xd6, 0xeb, 0xc6, 0x0e, 0xcf, 0xad, 0x08, 0x4e, 0xd7, 0xe3, 0x5d, 0x50, 0x1e, 0xb3,
+   0x5b, 0x23, 0x38, 0x34, 0x68, 0x46, 0x03, 0x8c, 0xdd, 0x9c, 0x7d, 0xa0, 0xcd, 0x1a, 0x41, 0x1c
+  };
+
+/* In all the *affine* intrinsics the argument b must be a compile-time constant so use macros */
+
+#if defined(SIMDE_GFNI_NATIVE) && defined(SIMDE_AVX512VL_NATIVE)
+  #define simde_mm_gf2p8affine_epi64_epi8(x, A, b) _mm_gf2p8affine_epi64_epi8(x, A, b)
+#else
+  SIMDE__FUNCTION_ATTRIBUTES
+  simde__m128i
+  simde_mm_gf2p8affine_epi64_epi8 (simde__m128i x, simde__m128i A, int b) {
+    #if defined(SIMDE_ARCH_X86_SSE4_1)
+      simde__m128i r, a, p;
+      const simde__m128i byteSelect = simde_mm_set_epi64x(UINT64_C(0x0101010101010101), 0);
+
+      a = simde_mm_shuffle_epi8(A, simde_mm_set_epi64x(UINT64_C(0x08090A0B0C0D0E0F), UINT64_C(0x0001020304050607)));
+      r = simde_mm_setzero_si128();
+
+      SIMDE__VECTORIZE
+      for (int i = 0 ; i < 8 ; i++) {
+	p = simde_mm_set1_epi16(HEDLEY_STATIC_CAST(uint16_t, simde_mm_movemask_epi8(a)));
+	p = simde_mm_shuffle_epi8(p, byteSelect);
+	p = simde_mm_xor_si128(r, p);
+	r = simde_mm_blendv_epi8(r, p, x);
+	a = simde_mm_slli_epi64(a, 1);
+	x = simde_mm_slli_epi64(x, 1);
+      }
+
+      return simde_mm_xor_si128(r, simde_mm_set1_epi8(HEDLEY_STATIC_CAST(uint8_t, b)));
+    #else
+      simde__m128i_private
+	r_,
+	x_ = simde__m128i_to_private(x),
+	A_ = simde__m128i_to_private(A);
+      const uint64_t ones = UINT64_C(0x0101010101010101);
+      const uint64_t mask = UINT64_C(0x0102040810204080);
+      uint64_t q;
+
+      SIMDE__VECTORIZE
+      for (size_t i = 0 ; i < (sizeof(r_.u8) / sizeof(r_.u8[0])) ; i++) {
+	q = HEDLEY_STATIC_CAST(uint64_t, x_.u8[i]) * ones;
+	q &= A_.u64[i / 8];
+	q ^= q >> 4;
+	q ^= q >> 2;
+	q ^= q >> 1;
+	q &= ones;
+	q *= 255;
+	q &= mask;
+	q |= q >> 32;
+	q |= q >> 16;
+	q |= q >> 8;
+	q ^= b;
+	r_.u8[i] = HEDLEY_STATIC_CAST(uint8_t, q);
+      }
+
+      return simde__m128i_from_private(r_);
+    #endif
+  }
+  #if defined(SIMDE_GFNI_ENABLE_NATIVE_ALIASES)
+    #define _mm_gf2p8affine_epi64_epi8(x, A, b) simde_mm_gf2p8affine_epi64_epi8(x, A, b)
+  #endif
+#endif
+
+#if defined(SIMDE_GFNI_NATIVE) && defined(SIMDE_AVX512VL_NATIVE)
+  #define simde_mm256_gf2p8affine_epi64_epi8(x, A, b) _mm256_gf2p8affine_epi64_epi8(x, A, b)
+#else
+  SIMDE__FUNCTION_ATTRIBUTES
+  simde__m256i
+  simde_mm256_gf2p8affine_epi64_epi8 (simde__m256i x, simde__m256i A, int b) {
+    simde__m256i_private
+      r_,
+      x_ = simde__m256i_to_private(x),
+      A_ = simde__m256i_to_private(A);
+
+    SIMDE__VECTORIZE
+    for (size_t i = 0 ; i < (sizeof(r_.m128i) / sizeof(r_.m128i[0])) ; i++) {
+      r_.m128i[i] = simde_mm_gf2p8affine_epi64_epi8(x_.m128i[i], A_.m128i[i], b);
```

Comment:
@nemequ Does `SIMDE__VECTORIZE` make sense here when the `simde_mm_gf2p8affine_epi64_epi8` function is so complex and there are only 2 slots?

Reply:
Thanks for the reviews.
In any case I shall have to change this to use macros as these intrinsics need the b argument to be a compile-time constant.

Reply:
Yes, it makes sense.  `SIMDE__VECTORIZE` is largely aspirational; we're just telling the compiler "if you can figure out how to vectorize this, please do!  There are no loop carried dependencies, so throw caution to the wind!"

Even if it doesn't do any good today, it's possible that in the future compilers will get smarter and it will help, so I'd rather leave it in.